command/job_restart.go

```diff
@@ -0,0 +1,1185 @@
+
+  -reschedule
+    If set, allocations are stopped and rescheduled instead of restarted
+    in-place. Since the group is not modified the restart does not create a new
+    deployment, and so values defined in 'update' blocks, such as
+    'max_parallel', are not taken into account. This option cannot be used with
+    '-task'.
+
+  -task=<task-name>
+    Specify the task to restart. Can be specified multiple times. If groups are
+    also specified the task must exist in at least one of them. If no task is
+    set only tasks that are currently running are restarted. For example,
+    non-sidecar tasks that already ran are not restarted unless '-all-tasks' is
+    used instead. This option cannot be used with '-all-tasks' or
+    '-reschedule'.
+
+  -yes
+    Automatic yes to prompts. If set, the command automatically restarts
+    multi-region jobs only in the region targeted by the command, ignores batch
+    errors, and automatically proceeds with the remaining batches without
+    waiting. Use '-on-error' and '-batch-wait' to adjust these behaviors.
+
+  -verbose
+    Display full information.
+`
+	return strings.TrimSpace(helpText)
+}
+
+func (c *JobRestartCommand) Synopsis() string {
+	return "Restart or reschedule allocations for a job"
+}
+
+func (c *JobRestartCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-all-tasks":         complete.PredictNothing,
+			"-batch-size":        complete.PredictAnything,
+			"-batch-wait":        complete.PredictAnything,
+			"-no-shutdown-delay": complete.PredictNothing,
+			"-on-error":          complete.PredictSet(jobRestartOnErrorAsk, jobRestartOnErrorFail),
+			"-reschedule":        complete.PredictNothing,
+			"-task":              complete.PredictAnything,
+			"-yes":               complete.PredictNothing,
+			"-verbose":           complete.PredictNothing,
+		})
+}
+
+func (c *JobRestartCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictFunc(func(a complete.Args) []string {
+		client, err := c.Meta.Client()
+		if err != nil {
+			return nil
+		}
+
+		resp, _, err := client.Search().PrefixSearch(a.Last, contexts.Jobs, nil)
+		if err != nil {
+			return []string{}
+		}
+		return resp.Matches[contexts.Jobs]
+	})
+}
+
+func (c *JobRestartCommand) Name() string { return "job restart" }
+
+func (c *JobRestartCommand) Run(args []string) int {
+	// Parse and validate command line arguments.
+	code, err := c.parseAndValidate(args)
+	if err != nil {
+		c.Ui.Error(err.Error())
+		c.Ui.Error(commandErrorText(c))
+		return code
+	}
+	if code != 0 {
+		return code
+	}
+
+	c.client, err = c.Meta.Client()
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error initializing client: %v", err))
+		return 1
+	}
+
+	// Use prefix matching to find job.
+	job, err := c.JobByPrefix(c.client, c.jobID, nil)
+	if err != nil {
+		c.Ui.Error(err.Error())
+		return 1
+	}
```

Comment:
During testing I discovered that passing `-reschedule` while there's an ongoing deployment pretty much always results in this command hanging. I used a job with a canary and manual promotion in its update block.

First attempt was running the restart while the deployment was waiting for manual promotion:

```
$ nomad job restart -reschedule -batch-wait=ask example
==> 2023-03-17T09:55:39-04:00: Restarting 7 allocations
    2023-03-17T09:55:39-04:00: Rescheduling allocation "3958e815" for group "group"
^C==> 2023-03-17T09:57:36-04:00: Restart interrupted, no more allocations will be restarted.
                               Are you sure you want to stop the restart process? [y/N]
    2023-03-17T09:57:39-04:00: Still waiting for allocation "3958e815" to be replaced
    2023-03-17T09:58:39-04:00: Still waiting for allocation "3958e815" to be replaced

$ nomad job allocs example
ID        Node ID   Task Group  Version  Desired  Status    Created    Modified
c7bfb048  22e1813e  group       1        run      running   52s ago    21s ago
3958e815  22e1813e  group       1        stop     complete  1m36s ago  46s ago
65110380  22e1813e  group       0        run      running   2m5s ago   1m48s ago
976fdbf3  22e1813e  group       0        run      running   2m13s ago  1m57s ago
a422b9db  22e1813e  group       0        run      running   2m22s ago  2m6s ago
a7c3945c  22e1813e  group       0        run      running   2m32s ago  2m16s ago
6fbff862  22e1813e  group       0        run      running   2m40s ago  2m24s ago
16472afa  22e1813e  group       0        run      running   4m44s ago  4m27s ago

$ nomad deployment list
ID        Job ID   Job Version  Status      Description
04edbf0c  example  1            running     Deployment is running but requires manual promotion
02cf29ff  example  0            successful  Deployment completed successfully
```

After canceling that and promoting the deployment, I deployed again. This time I waited until after promotion. In the example below, the allocation `1b86450c` was in a stopped state, having been replaced by the deployment:

```
$ nomad job restart -reschedule -batch-wait=ask example
==> 2023-03-17T10:03:37-04:00: Restarting 6 allocations
    2023-03-17T10:03:37-04:00: Rescheduling allocation "95f54f2c" for group "group"
    2023-03-17T10:03:37-04:00: Allocation "95f54f2c" replaced by "f254e087", waiting for "f254e087" to start running
    2023-03-17T10:03:43-04:00: Allocation "f254e087" is "running"
==> 2023-03-17T10:03:43-04:00: Proceed with the next batch? [Y/n/<wait duration>] 1s
==> 2023-03-17T10:03:59-04:00: Proceeding restarts with new wait time of 1s
==> 2023-03-17T10:03:59-04:00: Waiting 1s before restarting the next batch
    2023-03-17T10:04:00-04:00: Rescheduling allocation "57654725" for group "group"
    2023-03-17T10:04:00-04:00: Allocation "57654725" replaced by "7bce4be0", waiting for "7bce4be0" to start running
    2023-03-17T10:04:06-04:00: Allocation "7bce4be0" is "running"
==> 2023-03-17T10:04:06-04:00: Waiting 1s before restarting the next batch
    2023-03-17T10:04:07-04:00: Rescheduling allocation "1b86450c" for group "group"
    2023-03-17T10:05:07-04:00: Still waiting for allocation "1b86450c" to be replaced
    2023-03-17T10:06:07-04:00: Still waiting for allocation "1b86450c" to be replaced
```

My take on this is that we probably shouldn't support running `job restart` during an active deployment at all, because it makes it impossible for the deployment watcher to reason about what's going on. We could block that at this line by checking for active deployments and returning an error.

Reply:
Nice catch! We do the same thing for the autoscaler because messing with allocations during deployments is a recipe for disaster 😅 

I think there's a link missing in your comment, but I implemented the check in https://github.com/hashicorp/nomad/pull/16278/commits/069d2c6a06a052c34c4ea2913f509ce6e5257a4b before each restart. I also didn't gate it on `-reschedule` because I think even an in-place restart can be dangerous.

While working on this I realized that the core issue of the command getting blocked during a deployment was a bug in `monitorReplacementAlloc`. We need to keep following `NextAllocation` because, for example, if the first replacement fails Nomad will keep creating follow-up allocations. I fixed this in https://github.com/hashicorp/nomad/pull/16278/commits/fde22d65221dff71c5ac0a5dc1824691b73d03bf.

Reply:
> I think there's a link missing in your comment, but I implemented the check in 069d2c6 before each restart.

By "this line" I meant "here, right after `JobByPrefix`. But where you put it is much better, because that'll protect us from a time-of-check-time-of-use bug. :+1: 

> I also didn't gate it on `-reschedule` because I think even an in-place restart can be dangerous.

Good call

> While working on this I realized that the core issue of the command getting blocked during a deployment was a bug in `monitorReplacementAlloc`. We need to keep following `NextAllocation` because, for example, if the first replacement fails Nomad will keep creating follow-up allocations. I fixed this in https://github.com/hashicorp/nomad/commit/fde22d65221dff71c5ac0a5dc1824691b73d03bf.

Nice catch!